### PR TITLE
tikv-server: split toggle and generating memory profile

### PR DESCRIPTION
**Pull request by BusyJay**
So we can enable and take a profile whenever we want.

@siddontang @zhangjinpeng1987 PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
IMO, using other new two signals seems complex, and we will remove them later.

Maybe now we can enable profiling in environment directly and use SIGUSR2 to dump the profile. We can use sample to reduce the profile hurting for the peformance.

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/bin/profiling.rs [hunk 1]
```diff
@@ -42,66 +38,44 @@ mod imp {
...
-                          -> Option<JoinHandle<()>> {
-        if flag.load(Ordering::SeqCst) {
-            warn!("last memory profiling has not finished yet.");
-            return None;
-        }
+    /// Dump the profile to the `path`.
+    ///
+    /// If `path` is `None`, will dump it in the working directory with a auto-generated name.
+    pub fn dump_prof(path: Option<&str>) {
+        let mut c_path = DumpPathGuard::from_cstring(path.map(|p| CString::new(p).unwrap()));
```
Comment (siddontang) [inline src/bin/profiling.rs]: need to check active here?

File: src/bin/profiling.rs [hunk 2]
```diff
@@ -27,11 +23,11 @@ mod imp {
...
 
-    struct DumpFileNameGuard(Option<Vec<u8>>);
+    struct DumpPathGuard(Option<Vec<u8>>);
 
-    impl DumpFileNameGuard {
-        fn from_cstring(s: Option<CString>) -> DumpFileNameGuard {
-            DumpFileNameGuard(s.map(|s| s.into_bytes_with_nul()))
+    impl DumpPathGuard {
+        fn from_cstring(s: Option<CString>) -> DumpPathGuard {
+            DumpPathGuard(s.map(|s| s.into_bytes_with_nul()))
```
Comment (overvenus) [inline src/bin/profiling.rs]: What if `s` is `Some("")`?

Comment (BusyJay) [inline src/bin/profiling.rs]: It will be converted safely.

Comment (overvenus) [inline src/bin/profiling.rs]: What is the name of the dump file if `s` is `Some("")`, working directory + auto-generated name ?

Comment (BusyJay) [inline src/bin/profiling.rs]: No, it will failed and print a log.

File: src/bin/signal_handler.rs [hunk 3]
```diff
@@ -14,28 +14,28 @@
...
 
     use rocksdb::DB;
+    use libc;
 
     use tikv::server::Msg;
     use tikv::util::transport::SendCh;
     use prometheus::{self, Encoder, TextEncoder};
 
+    // Real-time signals
+    const TOGGLE_PROF_SIG: libc::c_int = 40;
```
Comment (hhkbp2) [inline src/bin/signal_handler.rs]: Is there any plan to switch to RPC like HTTP/GRPC later? It's handy to use signal for notify, but it's not the best way to add more and more signals for all that kind of things.

Comment (BusyJay) [inline src/bin/signal_handler.rs]: Yes, we thought about it. But rust grpc is not stable enough, so signal is still used here.

**Comment by hhkbp2** [comment]
LGTM

File: src/bin/signal_handler.rs (hunk 3 above)
Comment (siddontang) [inline src/bin/signal_handler.rs]: please add a TODO that we will remove this later.

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
@BusyJay 

CI failed 

```
 Compiling tikv v0.0.1 (file:///home/ubuntu/tikv)
$<2>error$<2>$<2>: unused import: `std::sync::Arc`$<2>
   $<2>$<2>--> $<2>src/bin/profiling.rs:110:9$<2>
    $<2>$<2>|$<2>
$<2>110$<2> $<2>$<2>| $<2>    use std::sync::Arc;$<2>
    $<2>$<2>| $<2>        $<2>$<2>^^^^^^^^^^^^^^$<2>
    $<2>$<2>|$<2>
    $<2>$<2>= $<2>$<2>note$<2>: `-D unused-imports` implied by `-D warnings`$<2>
```

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**